### PR TITLE
Fix reconstruction of CSI volumes

### LINKED ISSUE #117745: kubelet (1.27.1) fails to mount a hostPath volume for a static pod
### What happened?

This is a bit hard to reproduce, as it happens in a very specific setup, the problems only happens with `kubelet` 1.27.1, downgrading to 1.26.4 fixes the problem.

In a single-node cluster controlplane pods run as static pods, Rook/Ceph is deployed, and some PVs/PVCs backed by Ceph.

After a unclean shutdown related to some PVs backed by Ceph not being unmounted (this is not a Kubernetes), the `kubelet` which comes up after a reboot doesn't start controlplane static pods, which essentially blocks every other pod from starting (as API server is done).

After a reboot, kubelet is running, but it keeps reporting an error about static pods like:

```
172.20.0.2: {"ts":1683110056228.5916,"caller":"kubelet/kubelet.go:1875","msg":"Unable to attach or mount volumes for pod; skipping pod","pod":{"name":"kube-apiserver-talos-default-controlplane-1","namespace":"kube-system"},"err":"unmounted volumes=[audit secrets config], unattached volumes=[], failed to process volumes=[]: timed out waiting for the condition"}
```

All of the volumes are `hostPath` volumes in a static pod:

```yaml
        volumes:
            - hostPath:
                path: /system/secrets/kubernetes/kube-apiserver
              name: secrets
            - hostPath:
                path: /system/config/kubernetes/kube-apiserver
              name: config
            - hostPath:
                path: /var/log/audit/kube
              name: audit
```

The kubelet is stuck in this state "forever", I tested at least for 8 hours.

If I downgrade the kubelet (without machine reboot) to 1.26.4, it immediately starts the static pods bringing up the node online.

If I try same scenario with reboot with kubelet 1.26.4, the problem doesn't occur either.

### What did you expect to happen?

Kubelet should run static pods after a reboot.

### How can we reproduce it (as minimally and precisely as possible)?

It's a bit complicated set of workloads, but it's 100% reproducible. I could provide more logs if needed.

### Anything else we need to know?

/sig node

### Kubernetes version

<details>
1.27.1
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

Talos Linux v1.5.0-alpha

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>
containerd 1.6.20
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by jsafrane**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
During CSI volume reconstruction it's not possible to tell, if the volume is attachable or not - CSIDriver instance may not be available, because kubelet may not have connection to the API server at that time.
 
Therefore remove all calls to `FindAttachablePluginBySpec` here:

* Reorder `reconstructVolume` not to call `FindAttachablePluginBySpec` for CSI volumes. It's just a simple refactoring.
* Add *uncertain attach-abilty* state during reconstruction, so volume reconstruction can finish without an API server.  When the API server becomes available, then set attach-ability based on CSIDriver instances and allow unmounting of volumes.

#### Which issue(s) this PR fixes:

Fixes: #117745

#### Special notes for your reviewer:

~WIP:~
* ~PR for discussion~
* ~missing unit tests~ Unit tests added
* E2e test added in https://github.com/kubernetes/kubernetes/pull/118097

#### Does this PR introduce a user-facing change?

```release-note
Fixed kubelet startup getting stuck with `NewVolumeManagerReconstruction` feature enabled and a CSI volume present in /var/lib/kubelet/pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by jsafrane** [comment]
Added unit test and removed WIP:

**Comment by jingxu97** [comment]
@jsafrane would you mind adding some more details about the issue and fix?

**Comment by jsafrane** [comment]
@jingxu97, there are details in the issue, https://github.com/kubernetes/kubernetes/issues/117745#issuecomment-1534540981 + https://github.com/kubernetes/kubernetes/issues/117745#issuecomment-1534837217

**Comment by jsafrane** [comment]
And  I added some more details in the PR description. Check also individual commits - they should be relatively short.

**Comment by jsafrane** [comment]
I added a new e2e job, `pull-kubernetes-e2e-storage-kind-disruptive`. It runs kube-apiserver as a static pod and tries to reproduce https://github.com/kubernetes/kubernetes/issues/117745#issuecomment-1534540981.

* With `NewVolumeManagerReconstruction` feature enabled without any change, it fails: https://github.com/kubernetes/kubernetes/pull/118410
* With this PR + `NewVolumeManagerReconstruction` feature enabled  it succeeds: https://github.com/kubernetes/kubernetes/pull/118506

So, the issue should be fixed by this PR. I also tested manually with kube-apiserver as a static pod.

I'm proposing to review + merge this PR (or even https://github.com/kubernetes/kubernetes/pull/118506), let it soak in CI for a while and ask people who hit https://github.com/kubernetes/kubernetes/issues/117745 to test with master + `NewVolumeManagerReconstruction` enabled.

**Comment by jsafrane** [comment]
/retest
CPU quota flake

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go [hunk 1]
```diff
@@ -361,7 +375,20 @@ type mountedPod struct {
 func (asw *actualStateOfWorld) MarkVolumeAsAttached(
 	logger klog.Logger,
 	volumeName v1.UniqueVolumeName, volumeSpec *volume.Spec, _ types.NodeName, devicePath string) error {
-	return asw.addVolume(volumeName, volumeSpec, devicePath)
+
+	pluginIsAttachable := volumeAttachabilityFalse
+	if attachablePlugin, err := asw.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec); err == nil && attachablePlugin != nil {
```
Comment (msau42) [inline pkg/kubelet/volumemanager/cache/actual_state_of_world.go:380]: Is there any concern of marking the volume attachability as false if there's a temporary error?

File: pkg/kubelet/volumemanager/reconciler/reconstruct_common.go [hunk 2]
```diff
@@ -236,17 +236,25 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (rvolume *reconstructe
...
 	// the plugin supports attachable volumes.
-	attachablePlugin, err := rc.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec)
-	if err != nil {
-		return nil, err
-	}
 	deviceMountablePlugin, err := rc.volumePluginMgr.FindDeviceMountablePluginBySpec(volumeSpec)
 	if err != nil {
 		return nil, err
 	}
+	needsNameFromSpec := deviceMountablePlugin != nil
```
Comment (gnufied) [inline pkg/kubelet/volumemanager/reconciler/reconstruct_common.go:243]: Can we document why we need name from spec if volume plugin is attachable or mountable?

File: pkg/kubelet/volumemanager/reconciler/reconstruct_common.go [hunk 3]
```diff
@@ -236,17 +236,25 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (rvolume *reconstructe
...
+		// Check attach-ability of a volume only as a fallback to avoid calling
+		// FindAttachablePluginBySpec for CSI volumes - it needs a connection to the API server,
+		// but it may not be available at this stage of kubelet startup.
+		// All CSI volumes are device-mountable, so they won't reach this code.
+		attachablePlugin, err := rc.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec)
+		if err != nil {
+			return nil, err
+		}
+		needsNameFromSpec = attachablePlugin != nil
+	}
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/kubelet/volumemanager/reconciler/reconstruct_common.go:254]: Yes, this particular fix (the fist commit) just works around calling `FindAttachablePluginBySpec` for CSI volumes. CSI volumes are always device mountable, so they won't reach `FindAttachablePluginBySpec` here.

File: pkg/kubelet/volumemanager/reconciler/reconstruct_common.go (hunk 2 above)
Comment (jsafrane) [reply pkg/kubelet/volumemanager/reconciler/reconstruct_common.go:243]: I copied comment from DSWP: https://github.com/kubernetes/kubernetes/blob/1f2d49972cc541606b7a2a902271aa293b836d17/pkg/kubelet/volumemanager/cache/desired_state_of_world.go#L279-L283
The unique volume name must be reconstructed the same way.

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go [hunk 4]
```diff
@@ -361,7 +375,20 @@ type mountedPod struct {
...
+		pluginIsAttachable = volumeAttachabilityTrue
+	}
+
+	return asw.addVolume(volumeName, volumeSpec, devicePath, pluginIsAttachable)
+}
+
+func (asw *actualStateOfWorld) AddReconstructedVolume(
+	logger klog.Logger,
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go:388]: Removed

File: pkg/kubelet/volumemanager/cache/actual_state_of_world_test.go [hunk 5]
```diff
@@ -525,6 +525,54 @@ func TestActualStateOfWorld_FoundDuringReconstruction(t *testing.T) {
 				return nil
 			},
 		},
+		{
+			name: "uncertain attachability is resolved to attachable",
```
[the remark this answers is not in the document]
Comment (gnufied) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world_test.go:529]: So when we first build ASOW during reconsturction without API server connectivity we can't know atttachability of CSI volumes and hence "attachable" property is uncertain.

So reconciler continuously tries to fix that state and assuming once kubelet is able to talk to api-server uncertain attachability will be resolved.

File: pkg/kubelet/volumemanager/cache/actual_state_of_world.go [hunk 6]
```diff
@@ -169,14 +169,20 @@ type ActualStateOfWorld interface {
 	GetAttachedVolumes() []AttachedVolume
 
 	// SyncReconstructedVolume check the volume.outerVolumeSpecName in asw and
-	// the one populated from dsw , if they do not match, update this field from the value from dsw.
+	// the one populated from dsw, if they do not match, update this field from the value from dsw.
 	SyncReconstructedVolume(volumeName v1.UniqueVolumeName, podName volumetypes.UniquePodName, outerVolumeSpecName string)
 
+	// Add the specified volume to ASW as uncertainly attached.
+	AddReconstructedVolume(volumeName v1.UniqueVolumeName, volumeSpec *volume.Spec, nodeName types.NodeName, devicePath string) error
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/kubelet/volumemanager/cache/actual_state_of_world.go:176]: Renamed to `AddAttachUncertainReconstructedVolume`

**Comment by jsafrane** [comment]
I also renamed `volumesNeedDevicePath` to `volumesNeedUpdateFromNodeStatus` and `updateReconstructedFromAPIServer()` to `updateReconstructedFromNodeStatus` to reflect better what they actually do - see the last 2 commits.

**Pull request merged**